### DISCUSSION TITLE
Enabling Auto-Instrumentation via AutoConfiguredOpenTelemetrySdk rather than Java Agent

### DISCUSSION THREAD

**Discussion by paychex-ssmithrand** [open-telemetry/opentelemetry-java#5224] [Q&A]
I don't want to attach the Java agent jar to services, so I went down the route of setting up the SDK programmatically in a configuration file. Is there a way to add instrumentation libraries to the SDK in this fashion? Does `AutoConfiguredOpenTelemetrySdk` use the same environment variables as the agent to determine which libraries to instrument?
ie. `OTEL_INSTRUMENTATION_SPRING_MVC_ENABLED=true` should enable Spring MVC auto instrumentation for the agent.

I know there is a Spring starter module that will set up the interceptor, etc for requests/responses, but that starter does not include Spring Data, or non-http libraries to be instrumented.

EDIT: Doing some more digging around the SDK and related files, would this be done via a `TracerProvider` and passing in an auto-instrumentation class name?

Thanks!

**Comment by jkwatson** [comment]
No, there is no way to wire-up instrumentation via our autoconfigure module. That is only for configuring the SDK itself. 

It's an interesting idea, but I'm not sure exactly how it would work. We'd need to create a new SPI that instrumentation could implement. But, I don't know how that instrumentation would get attached to anything automatically.

**Reply by paychex-ssmithrand** [reply]
Thanks @jkwatson. Just to ensure we are on the same page on what I'd like to accomplish, I want to do something like this:

```
AutoConfiguredOpenTelemetrySdk configuredSdk = AutoConfiguredOpenTelemetrySdk.builder()
        .addPropertiesCustomizer((configProperties)->{
          HashMap<String,String> config = new HashMap<String,String>() {{
            put("otel.instrumentation.spring-data.enabled","true"); //Add a specific list of instrumentations I want auto-instrumented.
          }};
          return config;
        })
        .build();
```
I guess I'm curious how these properties work for the agent implementation but not for programmatic SDK implementation.

**Reply by jkwatson** [reply]
Well, the SDK doesn't know anything about any instrumentation libraries, or loading them, so I don't think this would do anything useful. It would have to involve something that the instrumentation could implement in order to set itself up. Even then, each instrumentation does something different (like grpc instrumentation creating interceptors, or http client instrumentation providing a wrapper around the client, etc). So, I don't think there's any simple approach to this right now. It would require a lot more than just setting some properties. :)

**Comment by trask** [comment]
@paychex-ssmithrand you might find this interesting? https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/runtime-attach
